sdks/java/io/jdbc/src/main/java/org/apache/beam/sdk/io/jdbc/JdbcIO.java

```diff
@@ -0,0 +1,492 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.sdk.io.jdbc;
+
+import com.google.common.base.Preconditions;
+
+import java.io.Serializable;
+import java.sql.Connection;
+import java.sql.PreparedStatement;
+import java.sql.ResultSet;
+
+import javax.annotation.Nullable;
+import javax.sql.DataSource;
+
+import org.apache.beam.sdk.transforms.Create;
+import org.apache.beam.sdk.transforms.DoFn;
+import org.apache.beam.sdk.transforms.PTransform;
+import org.apache.beam.sdk.transforms.ParDo;
+import org.apache.beam.sdk.transforms.display.DisplayData;
+import org.apache.beam.sdk.values.PBegin;
+import org.apache.beam.sdk.values.PCollection;
+import org.apache.beam.sdk.values.PDone;
+import org.apache.commons.dbcp2.BasicDataSource;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * <p>IO to read and write data on JDBC.</p>
+ * <p>
+ * <h3>Reading from JDBC datasource</h3>
+ * <p>
+ * JdbcIO source returns a bounded collection of {@code T} as a
+ * {@code PCollection<T>}. T is the type returned by the provided {@link RowMapper}.
+ * </p>
+ * <p>
+ * To configure the JDBC source, you have to provide a {@link DataSource}. The username and
+ * password to connect to the database are optionals. The following example illustrates how to
+ * configure a JDBC source:
+ * </p>
+ * <pre>
+ *   {@code
+ *
+ * pipeline.apply(JdbcIO.read()
+ *   .withDataSource(myDataSource)
+ *   .withQuery("select id,name from Person")
+ *   .withRowMapper(new JdbcIO.RowMapper<KV<Integer, String>>() {
+ *     public KV<Integer, String> mapRow(ResultSet resultSet) throws Exception {
+ *       KV<Integer, String> kv = KV.of(resultSet.getInt(1), resultSet.getString(2));
+ *       return kv;
+ *     }
+ *   })
+ *
+ *   }
+ * </pre>
+ * <h3>Writing to JDBC datasource</h3>
+ * <p>
+ * JDBC sink supports writing records into a database. It writes a {@link PCollection} to the
+ * database by converting each T into a {@link PreparedStatement} via a user-provided
+ * {@link PreparedStatementSetter}.
+ * </p>
+ * <p>
+ * Like the source, to configure JDBC sink, you have to provide a datasource. For instance:
+ * </p>
+ * <pre>
+ *   {@code
+ *
+ * pipeline
+ *   .apply(...)
+ *   .apply(JdbcIO.write()
+ *      .withDataSource(myDataSource)
+ *      .withQuery("insert into Person values(?, ?)")
+ *      .withPreparedStatementSetter(new JdbcIO.PreparedStatementSetter<KV<Integer, String>>() {
+ *        public void setParameters(KV<Integer, String> element, PreparedStatement query) {
+ *          // use the PCollection element to set parameters of the SQL query used to insert
+ *          // in the database
+ *          // for instance:
+ *          query.setInt(1, kv.getKey());
+ *          query.setString(2, kv.getValue());
+ *        }
+ *      })
+ *
+ *   }
```

Comment:
Add a note that the statements may be executed multiple times in case of retries, due to failures and such, so normally you should use https://en.wikipedia.org/wiki/Merge_(SQL) instead of INSERT unless you're fine with duplicate rows. The example query and the unit test should probably be updated as well.